packages/amplify-provider-awscloudformation/src/pre-push-cfn-processor/cfn-pre-processor.ts

```diff
@@ -24,3 +38,155 @@ export async function preProcessCFNTemplate(filePath: string): Promise<string> {
+
+async function addCustomPoliciesToCFNTemplate(
+  service: string,
+  category: string,
+  customPolicies: CustomIAMPolicies,
+  cfnTemplate: Template,
+  resourceName: string
+  ) {
+  let customExecutionPolicy;
+  if (service === 'Lambda') {
+    customExecutionPolicy = customExecutionPolicyForFunction;
+  }
+  if (service === 'ElasticContainer') {
+    const roleName = cfnTemplate.Resources.TaskDefinition.Properties.ExecutionRoleArn['Fn::GetAtt'][0];
+    customExecutionPolicy = customExecutionPolicyForContainer;
+    const role = Fn.Ref(roleName);
+    customExecutionPolicy.Properties.Roles.push(role);
+  }
+
+  warnWildcardCustomPolicies(customPolicies, resourceName);
+
+  for (const customPolicy of customPolicies) {
+    validateCustomPolicy(customPolicy, category, resourceName);
+    const policyWithEnv = replaceEnvForCustomPolicies(customPolicy);
+    if (!policyWithEnv.Effect) policyWithEnv.Effect = 'Allow';
+    customExecutionPolicy.Properties.PolicyDocument.Statement.push(policyWithEnv);
+  }
+
+  if (service === 'Lambda') {
+    cfnTemplate.Resources.CustomLambdaExecutionPolicy = customExecutionPolicy;
+  }
+  if (service === 'ElasticContainer') {
+    cfnTemplate.Resources.CustomExecutionPolicyForContainer = customExecutionPolicy;
+  }
+
+  return cfnTemplate;
+}
+
+
+//validate the format of actions and ARNs for custom IAM policies
+function validateCustomPolicy(customPolicy: CustomIAMPolicy, category: string, resourceName: string) {
+  const resources = customPolicy.Resource;
+  const actions = customPolicy.Action;
+  const resourceRegex = new RegExp('arn:(aws[a-zA-Z0-9-]*):([a-zA-Z0-9\\-])+:([a-z]{2}(-gov)?-[a-z]+-\\d{1})?:(\\d{12})?:(.*)');
+  const actionRegex = new RegExp('[a-zA-Z0-9]+:[a-z|A-Z|0-9|*]+');
+  const wrongResourcesRegex = [];
+  const wrongActionsRegex = [];
+  let errorMessage = '';
+
+  for (const resource of resources) {
+    if (!(resourceRegex.test(resource) || resource === '*')) {
+      wrongResourcesRegex.push(resource);
+    }
+  }
+
+  for (const action of actions) {
+    if (!actionRegex.test(action)) {
+      wrongActionsRegex.push(action);
+    }
+  }
+
+  const customPoliciesPath = pathManager.getCustomPoliciesPath(category, resourceName);
+
+  if (wrongResourcesRegex.length > 0) {
+    errorMessage += `Invalid custom IAM policy for ${resourceName}. Incorrect "Resource": ${wrongResourcesRegex.toString()}\n Edit ${customPoliciesPath} to fix`;
+  }
+  if (wrongActionsRegex.length > 0) {
+    errorMessage += `Invalid custom IAM policy for ${resourceName}. Incorrect "Action": ${wrongActionsRegex.toString()}\n Edit ${customPoliciesPath} to fix`;
+  }
+
+  if (errorMessage.length > 0) {
+    throw new CustomPoliciesFormatError(errorMessage);
+  }
+}
+
+function doCustomPoliciesExist(customPolicies: CustomIAMPolicies) : Boolean | PromiseLike<any> {
+  if (!customPolicies || customPolicies.length === 0) {
+    return false;
+  }
+
+  if (customPolicies.length === 1
```

Comment:
Yes, this is a good one. It will get an error later on the check validate schema. I am doing the validate exist just for the situation they never edit the file or just delete things they enter. I did think about I can do something check every object to see if it contains something. But it also makes sense this got caught by validate schema, for customers did edit the file. What do you think?

Reply:
oh, yeah I think it's fine since it will get caught by ajv